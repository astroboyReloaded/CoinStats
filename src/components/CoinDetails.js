import { useDispatch, useSelector } from 'react-redux';
import { useEffect } from 'react';
import { Link, useParams } from 'react-router-dom';
import parser from 'html-react-parser';
import {
  fetchCoinDetails,
  clearCoinDetails,
} from '../redux/coin-details/coinDetailsSlice';
import PriceConvertion from './PriceConvertion';
import '../styles/CoinDetails.css';
import useSetThisCoinInFilter from '../hooks/useSetThisCoinInFilter';

const CoinDetails = () => {
  const { id } = useParams();
  const {
    coinDetails,
    isLoading,
    error,
  } = useSelector(
    (state) => state.coinDetails,
  );

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCoinDetails(id));
    return () => {
      dispatch(clearCoinDetails());
    };
  }, [dispatch, id]);

  useSetThisCoinInFilter(coinDetails);

  const coin = coinDetails;

  const {
    symbol,
    name,
    description,
    image,
    market_data: marketData,
  } = coinDetails;

  return (
    <main className="coin-main">
      {isLoading && <p>...loading</p>}
      {error && <p>Ups! There was an error...</p>}
      {coin !== 'Coin Details' && (
        <article>
          <header className="coin-header">
            <Link className="arrow-back" to="/" />
            <img className="header-coin-image" src={image.small} alt={name} />
            <h1 className="coin-name">{name}</h1>
            <h2 className="coin-symbol">{`(${symbol?.toUpperCase()})`}</h2>
          </header>
          <main className="coin-main-section">
            <section>
              <data
                className="coin-details-price"
                value={marketData.current_price.usd}
              >
                {`$${marketData.current_price.usd.toLocaleString()}`}
              </data>
              <data className="mrkt-cap-chng-perc">{`${marketData.market_cap_change_percentage_24h}%`}</data>
              <PriceConvertion />
            </section>
            <section className="priceChange-percentages">
              <table>
                <thead>
                  <tr>
                    <th>24H</th>
                    <th>7D</th>
                    <th>14D</th>
                    <th>30D</th>
                    <th>60D</th>
                    <th>1Y</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>{marketData.price_change_percentage_24h}</td>
                    <td>{marketData.price_change_percentage_7d}</td>
                    <td>{marketData.price_change_percentage_14d}</td>
                    <td>{marketData.price_change_percentage_30d}</td>
                    <td>{marketData.price_change_percentage_60d}</td>
                    <td>{marketData.price_change_percentage_1y}</td>
                  </tr>
                </tbody>
              </table>
            </section>
            <section>
              <table className="coin-MainData">
                <thead>
                  <tr className="coin-MainData-Titles">
                    <th>Market Cap Rank</th>
                    <th>Market Cap</th>
                    <th>Fully Diluted Valuation</th>
                    <th>Trading Volume</th>
                    <th>2H High</th>
                    <th>2H Low</th>
                    <th>Circulating Supply</th>
                    <th>Total Supply</th>
                    <th>Max Supply</th>
                    <th>All-Time High</th>
                    <th>All-Time Low</th>
                  </tr>
                </thead>
                <tbody>
                  <tr className="coin-MainData-Values">
                    <td>{marketData.market_cap_rank}</td>
                    <td>{marketData.market_cap?.usd}</td>
                    <td>{marketData.fully_diluted_valuation?.usd}</td>
                    <td>{marketData.total_volume?.usd}</td>
                    <td>{marketData.high_24h?.usd}</td>
                    <td>{marketData.low_24h?.usd}</td>
                    <td>{marketData.circulating_supply}</td>
                    <td>{marketData.total_supply}</td>
                    <td>{marketData.max_supply}</td>
                    <td>{marketData.ath?.usd}</td>
                    <td>{marketData.atl?.usd}</td>
                  </tr>
                </tbody>
              </table>
            </section>
            <p className="coin-description">{parser(description.en)}</p>
          </main>
        </article>
      )}
    </main>
  );
};

export default CoinDetails;
