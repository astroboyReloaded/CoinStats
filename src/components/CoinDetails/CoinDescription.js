import { useSelector } from 'react-redux';
import parser from 'html-react-parser';

const CoinDescription = () => {
  const { description } = useSelector((state) => state.coinDetails.coinDetails);
  return (
    <article>
      <p className="coin-description">{parser(description)}</p>
    </article>
  );
};

export default CoinDescription;
