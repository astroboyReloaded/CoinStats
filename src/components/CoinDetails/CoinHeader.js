import { useSelector } from 'react-redux';
import { Link } from 'react-router-dom';

const CoinHeader = () => {
  const { name, symbol, image } = useSelector((state) => state.coinDetails.coinDetails);

  return (
    <header className="coin-header">
      <Link className="arrow-back" to="/" />
      <img className="header-coin-image" src={image} alt={name} />
      <h1 className="coin-name">{name}</h1>
      <h2 className="coin-symbol">{`(${symbol?.toUpperCase()})`}</h2>
    </header>
  );
};

export default CoinHeader;
