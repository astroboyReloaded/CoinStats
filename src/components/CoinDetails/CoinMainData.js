import useCoinMainData from '../../hooks/coinDetails-hooks/useCoinMainData';
import { genSymbol } from '../../hooks/helperFuncs';

const CoinMainData = () => {
  const coinMainData = useCoinMainData();

  if (coinMainData) {
    return (
      <section>
        <table className="coin-MainData">
          <thead className="con-min-data-thead-tbody">
            <tr className="coin-MainData-tr">
              {Object.keys(coinMainData).map((cmdKey) => (
                <th key={cmdKey} className="coinMD-th">{cmdKey}</th>
              ))}
            </tr>
          </thead>
          <tbody className="con-min-data-thead-tbody">
            <tr className="coin-MainData-tr">
              {Object.keys(coinMainData).map((cmdkey, i) => (
                <td key={cmdkey} className="coinMD-td">
                  {`${genSymbol(i)}${coinMainData[cmdkey]?.toLocaleString()}`}
                </td>
              ))}
            </tr>
          </tbody>
        </table>
      </section>
    );
  }
  return <span>Awaiting data...</span>;
};

export default CoinMainData;
