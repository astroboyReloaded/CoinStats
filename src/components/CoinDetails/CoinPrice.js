import { useSelector } from 'react-redux';

const CoinPrice = () => {
  const { currentPrice, marketData } = useSelector((store) => store.coinDetails.coinDetails);

  return (
    <section className="coin-price-section">
      <data
        className="coin-details-price"
        value={currentPrice.usd}
      >
        {`$${currentPrice.usd.toLocaleString()}`}
      </data>
      <data className={`mrkt-cap-chng-perc ${marketData.market_cap_change_percentage_24h > 0 ? 'green' : 'red'}`}>{`${marketData.market_cap_change_percentage_24h.toFixed(2).replace('-', '')}%`}</data>
    </section>
  );
};

export default CoinPrice;
