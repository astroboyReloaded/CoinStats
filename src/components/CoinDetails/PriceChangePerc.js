import useExtractPercs from '../../hooks/coinDetails-hooks/useExtractPercs';

const PriceChangePerc = () => {
  const {
    ready,
    priceChangePercs,
  } = useExtractPercs();

  if (ready) {
    return (
      <section className="priceChange-percentages">
        <table className="price-percentage-table">
          <thead>
            <tr className="price-perc-tr">
              {Object.keys(priceChangePercs).map((pcpKey) => <th key={pcpKey} className="price-percentage-th">{pcpKey}</th>)}
            </tr>
          </thead>
          <tbody>
            <tr className="price-perc-tr">
              {Object.keys(priceChangePercs).map((pcpKey) => <td key={Math.random()} className={`price-percentage-td ${priceChangePercs[pcpKey] > 0 ? 'green' : 'red'}`}>{`${priceChangePercs[pcpKey].toFixed(2).replace('-', '')}%`}</td>)}
            </tr>
          </tbody>
        </table>
      </section>
    );
  }
  return <span>awaiting data...</span>;
};

export default PriceChangePerc;
