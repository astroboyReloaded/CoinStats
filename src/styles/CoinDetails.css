.coin-main {
  padding: 0.3rem;
  font-family: Arial, Helvetica, sans-serif;
  color: gainsboro;
}

.coin-header {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.arrow-back {
  display: block;
  height: 1.8rem;
  width: 1.8rem;
  background: url('../../public/icons/arrow-left.png') 50% no-repeat;
  background-size: 18px;
}

.header-coin-image {
  width: 18px;
  height: 18px;
  margin-inline: 0.6rem;
}

.coin-name,
.coin-symbol {
  margin: 0;
  font-size: 1rem;
  font-weight: 300;
}

.coin-main-section {
  display: flex;
  flex-direction: column;
}

.coin-details-price {
  font-size: 2.1rem;
  width: fit-content;
}

.mrkt-cap-chng-perc {
  width: fit-content;
}

.coin-MainData {
  display: flex;
}

.coin-MainData-Titles,
.coin-MainData-Values {
  display: flex;
  flex-direction: column;
}

a {
  color: goldenrod;
}

