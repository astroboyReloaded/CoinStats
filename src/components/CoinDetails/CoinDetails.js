import { useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';
import useFetchCoinDetails from '../../hooks/coinDetails-hooks/useFetchCoinDetails';
import '../../styles/CoinDetails.css';
import CoinHeader from './CoinHeader';
import CoinPrice from './CoinPrice';
import PriceConvertion from './PriceConvertion/PriceConvertion';
import PriceChangePerc from './PriceChangePerc';
import CoinMainData from './CoinMainData';
import CoinDescription from './CoinDescription';
import Footer from '../Footer';
import Loading from '../Loading';
import PriceChart from './PriceChart';

const CoinDetails = () => {
  const { id } = useParams();
  const { isLoading, error, ready } = useSelector((state) => state.coinDetails);

  useFetchCoinDetails(id);

  if (isLoading) {
    return <Loading />;
  }
  if (ready) {
    return (
      <div className="coinContainer">
        <CoinHeader />
        <main className="coin-main-section">
          <CoinPrice />
          <PriceConvertion />
          <PriceChart coinID={id} />
          <PriceChangePerc />
          <CoinMainData />
          <CoinDescription />
        </main>
        <Footer />
      </div>
    );
  }
  return <p>{`Ups! There was an error... ${error}`}</p>;
};

export default CoinDetails;
