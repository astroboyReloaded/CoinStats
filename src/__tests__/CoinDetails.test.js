import renderer from 'react-test-renderer';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import CoinDetails from '../components/CoinDetails/CoinDetails.js';

it('renders CoinDetails', () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Routes>
          <Route path="/coin-details/:id" element={<CoinDetails />} />
        </Routes>
      </BrowserRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
