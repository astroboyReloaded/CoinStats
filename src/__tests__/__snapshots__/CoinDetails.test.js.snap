// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CoinDetails 1`] = `null`;
